### PR TITLE
Analysis/analysis model

### PR DESCRIPTION
**Pull request by kylesayrs**
This PR adds the AnalysisModel class which, given a model onnx, produces a json that specifies features about that model such as its sparsity, number of flops/int ops, and four block sparsity on a global and per-node basis.

### CODE REVIEW

**Comment by bfineran** [comment]
until #151 lands, going to point the base of this this PR to `analysis/helper-files`, so only the appropriate changes will be shown

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:1]: nit: file name should be changed to avoid upper case. if the upper case was a workaround for namespace conflict with the parent directory, we can rename it to `model_analysis` or (I much less prefer) `main.py`

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:92]: nit: let's call this operation counts

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:99]: we can leave these out the dicts provide better information than this with about the same effort of looking through

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:123]: let's rename to layers since `Node` is `ONNX` specific

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:18]: nit: just `numpy`

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:136]: for readability and since we are casting to a list anyways, I'd suggest switching this to a list comprehension with an `if` for filtering

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:169]: similar to above, calling sum directly and using a comprehension will be a bit cleaner and readable.  I do like the helper function, just not a big value add here

File: src/sparsezoo/analysis/Analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/Analysis.py:191]:
```suggestion
    @staticmethod
    def analyze_nodes(model_onnx: ModelProto) -> List[NodeAnalysis]:
```

**Review by bfineran — CHANGES_REQUESTED** [review]
looks good overall, great job getting started with pydantic and making use of your helpers from #151,
left some comments to get started will go through tests next week

**Comment by kylesayrs** [comment]
Still need to implement flops and ops on a per layer basis and in total

**Comment by kylesayrs** [comment]
Open to suggestions on how to deal with the functions in onnx_helpers.py. These functions were taken from sparseml and slightly modified. Make throw both onnx_helpers and helpers into a utils directory?

File: src/sparsezoo/analysis/analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/analysis.py:83]: let's rename this to `non_parameterized_operator_counts`

File: src/sparsezoo/analysis/analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/analysis.py:194]: is this repeated in the classmethod above? looks like we are running this function (and extract_node_shapes) twice

Comment (bfineran) [inline src/sparsezoo/analysis/analysis.py:194]: is this repeated in the classmethod above? looks like we are running this function (and extract_node_shapes) twice

File: src/sparsezoo/analysis/analysis.py
Comment (bfineran) [inline src/sparsezoo/analysis/analysis.py:148]: can this section be completed by using the `NodeAnalysis` list rather than re-doing here?

Comment (bfineran) [inline src/sparsezoo/analysis/analysis.py:148]: can this section be completed by using the `NodeAnalysis` list rather than re-doing here?

File: src/sparsezoo/analysis/helpers.py
Comment (bfineran) [inline src/sparsezoo/analysis/helpers.py:48]: nit: "private" functions should go at the bottom of the file after exported ones

Comment (bfineran) [inline src/sparsezoo/analysis/helpers.py:48]: nit: "private" functions should go at the bottom of the file after exported ones

File: src/sparsezoo/analysis/onnx_helpers.py
Comment (bfineran) [inline src/sparsezoo/analysis/onnx_helpers.py:18]:
```suggestion
NOTE: Adapted from sparseml/onnx/utils/helpers.py with minimal edits
```
:)

Comment (bfineran) [inline src/sparsezoo/analysis/onnx_helpers.py:18]:
```suggestion
NOTE: Adapted from sparseml/onnx/utils/helpers.py with minimal edits
```
:)

**Review by bfineran — CHANGES_REQUESTED** [review]
looks great overall, we'll need to do some sort of check to make sure onnxruntime is installed and either raise an error or silently omit flops calculations if not. let's also talk about how the tests are setup offline.

Curious to see what the plain text analysis will look like. let's go through that together and see if any adjustments need to be made accordingly

**Comment by kylesayrs** [comment]
This PR is ready for review. Notably I implemented get_num_dense_and_sparse_ops which uses node shapes and weights to determine approximately how many ops each node performs.

I ran into an issue where the node shapes inferred using previous code from sparseml/onnx/utils/helpers.py was giving None because of incomplete implementations in onnxruntime. I instead run inference on the model with inputs of all ones to get real output values which I then use to generate shapes. This probably needs a little more testing with models with dynamic shape.

**Comment by kylesayrs** [comment]
I've also reorganized the functions into separated files, if that makes them easier to read. The reorganization is in a commit on this branch: https://github.com/neuralmagic/sparsezoo/tree/analysis/chart-utils

**Review by bfineran — APPROVED** [review]
lgtm pending audit of analysis file

File: src/sparsezoo/analysis/analysis.py
Comment (dbogunowicz) [inline src/sparsezoo/analysis/analysis.py:100]:
```suggestion
        :param node_shapes: optional dictionary of node shapes. If not supplied,
```

File: src/sparsezoo/analysis/helpers.py
Comment (dbogunowicz) [inline src/sparsezoo/analysis/helpers.py:73]: Where do indexes 8 and 2 come from? (L73 and  L70)

File: src/sparsezoo/analysis/helpers.py
Comment (dbogunowicz) [inline src/sparsezoo/analysis/helpers.py:95]:
```suggestion
    :param node_shapes: optional dictionary of node shapes. If not supplied,
```

File: src/sparsezoo/analysis/helpers.py
Comment (dbogunowicz) [inline src/sparsezoo/analysis/helpers.py:97]:
```suggestion
    :param is_four_block_sparse: optional boolean indicating if this node is four
```

File: src/sparsezoo/analysis/helpers.py
Comment (dbogunowicz) [inline src/sparsezoo/analysis/helpers.py:116]: why do we split into two statements?
i.e.
why not `if node.op_type in list1 + list2` instead of `if node.op_type in list1 or list2`?

**Comment by dbogunowicz** [comment]
@kylesayrs couple of nits from my side, otherwise great job.

**Pull request merged**
